File: ts/src/hooks/codec.ts
import type { Capabilities } from "../capabilities";
import type { Input } from "./input";
import type { Decision } from "./decision";

/** Translates between a host's native hook wire format and the canonical Input/Decision types. */
export interface Codec {
  host(): string;
  /** Builds the host's native hook stdin envelope, as a JSON value. */
  encodeInput(input: Input): unknown;
  /** Parses a host's native hook stdin envelope (already-parsed JSON value). */
  decodeInput(raw: unknown): Input;
  /** Encodes a Decision into the host's native stdout/exit-code convention. */
  encodeDecision(d: Decision): { stdout: unknown; exit: number };
  /** Decodes a host's stdout/exit-code convention back into a Decision. */
  decodeDecision(stdout: unknown, exit: number): Decision;
  capabilities(): Capabilities;
}

File: py/src/axon/hooks/codec.py
"""The Codec protocol: translates between a host's native hook wire
format and the canonical Input/Decision types. Mirrors ts/src/hooks/codec.ts.
"""

from __future__ import annotations

from typing import Protocol

from axon.capabilities import Capabilities
from axon.hooks.decision import Decision
from axon.hooks.input import Input


class Codec(Protocol):
    def host(self) -> str: ...

    def encode_input(self, inp: Input) -> dict:
        """Builds the host's native hook stdin envelope, as a JSON value."""
        ...

    def decode_input(self, raw: dict) -> Input:
        """Parses a host's native hook stdin envelope (already-parsed JSON value)."""
        ...

    def encode_decision(self, d: Decision) -> tuple[object, int]:
        """Encodes a Decision into the host's native stdout/exit-code convention.

        Returns (stdout, exit).
        """
        ...

    def decode_decision(self, stdout: object, exit_code: int) -> Decision:
        """Decodes a host's stdout/exit-code convention back into a Decision."""
        ...

    def capabilities(self) -> Capabilities: ...
